Allow empty partition keys in views

nyh opened the pull request:
Although Cassandra generally does not allow empty strings as partition
keys (they are allowed as clustering keys!), it *does* allow empty strings
in regular columns to be indexed by a secondary index, or to become an
empty partition-key column in a materialized view. As noted in issues #9375
and #9364 and verified in a few xfailing cql-pytest tests, Scylla didn't
allow these cases - and this two-patch series fixes that.

The first patch fixes #9352 - an inconsistency in our tokenizer which
resulted in Scylla not being able to look up an empty key in the
sstable's index. Then, the second patch fixes the materialized views
implementation to not drop view rows with an empty-string partition key.
The fix for secondary index bug comes "for free" because it is based
on materialized views.

We already had xfailing test cases for empty strings in materialized
views and indexes, and after this series they begin to pass so the
"xfail" mark is removed. The series also adds additional test cases
that validate additional corner cases discovered during the debugging.

Fixes #9352
Fixes #9364
Fixes #9375

avikivity commented:
> > We translate token ranges to (low, high]. So (minimum_token, some_token] excludes minimum_token.
> 
> Are you sure about that? 

No.

> First, I see that minimum_token() isn't simply INT64_MIN, it actually has the value 0 (!) but with the "kind" before_all_keys. 

before_all_keys ignores the numeric value of the token. So this isn't compatible with Cassandra, since it uses a different token.

> Second, the function shard_of() (see dht/token.cc) specifically returns a number (shard_of_minimum_token(), or 0) for minimum_token so it is assigned to a specific shard. 

This is just to allow range scans not to jump around. It wasn't expected at the time of writing that it would be used for anything.

In fact, @haaawk and I want to remove token::_kind and replace its uses with infinite interval bounds.

> Third, as assignments to nodes go, I believe the token range between the very last ring token and the very first ring token does include minimum_token(), so minimum_token() is contained inside one (wrapping) token range. Isn't it?

Yes. But internally this is translated to nonwrapping intervals, so it may fall off the edge.

> 
> I have to admit that my tests all involved a single node, in cql-pytest. I will need to also add a dtest to get a better feeling that this minimum_token() thing really works in large clusters. Maybe it really doesn't work - I need to check it.
> 
> I agree that this special-case is ugly. I'm scared of trying to remove it - first because it's hard to know which iterations in Scylla which rely on being able to name the "first partition key" to start iterating on will stop working, and second because I'm worried we might regret this in the future if we ever need to import from Cassandra tables with the empty key (though as I noted earlier, today this isn't relevant - we don't import views and Cassandra doesn't allow empty keys in base tables).

It's already incompatible because you're using a different minimum token - one that must not have its value examined. 

Using 0x8000'0000'0000'0000 looks most compatible, but it threatens to confuse ring calculations.

nyh commented:
**NOTE TO SELF:**

In `sstable::set_first_and_last_keys() ` we have a sanity check that the first and last partition keys in an sstable are not empty, and if it is, we refuse to write the sstable. I'm worried that a materialized view or index whose first or last key is empty will therefore break. We need to write a test where we have only an empty key, or alternatively write just the empty key between "nodetool flush" of the table, and see that the sstable writing doesn't break. I am worried that I didn't fix this in my patches.

nyh commented:
Downgrading this PR to draft. I needs rebasing, and when I rebase it it doesn't work - apparently something changed in the validator code so my fix for them in the first patch somehow broken Scylla boot, failing to write sstables.

nyh commented:
Rebased, fixed a small bug (missing "return true"...) and this series now works again. Still haven't implemented the "NOTE TO SELF" from above, and still won't please Avi because he doesn't want me to keep the weird exception for the token of the empty partition.

nyh commented:
I pushed a version which fully works (and the note I left to myself above turned out to be a non-issue), so removing the "draft" flag.

However, please don't commit yet because @avikivity doesn't like this approach, and would like me, instead, to drop the pre-existing special-case for the token of the empty string (it should just be the normal formula, which will give it a token of 0).
I'll do this later, probably in a new pull-request.  We **can't** fix this incrementally, because as soon as this patch goes in, user may start having empty-string keys in their materialized views - and we can't change the token of those keys later or we break backward-compatibility.

So in the mean time, please don't merge this pull-request.

nyh commented:
Converted this issue to a draft again, because it is susceptible to another bug - https://github.com/scylladb/scylla/issues/10176.
I'll fix that bug in the context of the other PR, and probably this PR will remain a draft indefinitely until I close it.

nyh commented:
Converted this issue to a draft again, because it is susceptible to another bug - https://github.com/scylladb/scylla/issues/10176.
I'll fix that bug in the context of the other PR, and probably this PR will remain a draft indefinitely until I close it.

nyh commented:
Converted this issue to a draft again, because it is susceptible to another bug - https://github.com/scylladb/scylla/issues/10176.
I'll fix that bug in the context of the other PR, and probably this PR will remain a draft indefinitely until I close it.
